runtime/compiler/net/ServerStream.hpp

```diff
@@ -0,0 +1,238 @@
+
+      @return Returns a tuple of arguments sent by the client
+   */
+   template <typename ...T>
+   std::tuple<T...> read()
+      {
+      readBlocking(_cMsg);
+      if (_cMsg.type() == MessageType::compilationAbort)
+         throw StreamCancel(_cMsg.type());
+      // We are expecting the response type (_cMsg.type()) to be the same as the request type (_sMsg.type())
+      if (_cMsg.type() != _sMsg.type())
+         throw StreamMessageTypeMismatch(_sMsg.type(), _cMsg.type());
+         
+      return getArgs<T...>(_cMsg.mutable_data());
+      }
+
+   /**
+      @brief Function to read the compilation request from a client
+
+      This is the first type of message received after a connection is established.
+      The number and position of parameters in the template must match the
+      the one sent by the client. In order to ensure this, the client will embed
+      version information in the first message it sends after a connection is established.
+      The server will check whether its version matches the client's version and throw
+      `StreamVersionIncompatible` it is doesn't.
+
+      Exceptions thrown: StreamCancel, StreamVersionIncompatible,StreamMessageTypeMismatch
+ 
+      @return Returns a tuple with information sent by the client
+   */
+   template <typename... T>
+   std::tuple<T...> readCompileRequest()
+      {
+      readBlocking(_cMsg);
+      if (_cMsg.type() == JITServer::MessageType::clientTerminate)
+         {
+         uint64_t clientId = std::get<0>(getRecvData<uint64_t>());
+         throw JITServer::StreamCancel(_cMsg.type(), clientId);
+         }
+      if (_cMsg.version() != 0 && _cMsg.version() != getJITaaSVersion())
+         {
+         throw StreamVersionIncompatible(getJITaaSVersion(), _cMsg.version());
+         }
+      if (_cMsg.type() != JITServer::MessageType::compilationRequest)
+         {
+         throw JITServer::StreamMessageTypeMismatch(JITServer::MessageType::compilationRequest, _cMsg.type());
+         }
+      return getArgs<T...>(_cMsg.mutable_data());
+      }
+   
+   /**
+      @brief Extract the data from the received message and return it
+   */
+   template <typename... T>
+   std::tuple<T...> getRecvData()
+      {
+      return getArgs<T...>(_cMsg.mutable_data());
+      }
+
+   /**
+      @brief Function invoked by server when compilation is completed
+
+      This should be the last message sent by a server as a response to a compilation request.
+      It must include a `statusCode` which can be indicative of an error and could include a 
+      variable number of parameters with compilation artifacts (including the compiled body).
+   */
+   template <typename... T>
+   void finishCompilation(uint32_t statusCode, T... args)
+      {
+      try
+         {
+         write(MessageType::compilationCode, statusCode, args...);
+         }
+      catch (std::exception &e)
+         {
+         if (TR::Options::getVerboseOption(TR_VerboseJITaaS))
```

Comment:
Shouldn't this do something more than just log the exception?

Reply:
`finishCompilation()` is trigged on the server side. Failures are handled on the client side.

Reply:
If the write fails, does the client just time out?

Reply:
Yes, the client will timeout. Since the write failed, there is no way to inform the client that something wrong happened. The server will just close this connection and continue to wait for other connection requests.